packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,384 @@
+ * NodePart(N16); in this case the entire part is being disconnected, so we
+ * simply iterate all of N16's `_$disconnectableChildren` (D17,T18) and
+ * recursively run `setConnected` on them. Note that we only remove children
+ * from `_$disconnectableChildren` for the top-level values being disconnected
+ * on a clear; doing this bookkeeping lower in the tree is wasteful since it's
+ * all being thrown away.
+ *
+ * Example 3: If the LitElement containing the entire tree above becomes
+ * disconnected, it will run `nodePart.setConnected()` (which calls
+ * `nodePart._$setNodePartConnected()` if it exists); in this case, we
+ * recursively run `setConnected()` over the entire tree, without removing any
+ * children from `_$disconnectableChildren`, since this tree is required to
+ * re-connect the tree, which does the same operation, simply passing
+ * `isConnectd: true` down the tree, signaling which callback to run.
+ */
+
+import {
+  Directive,
+  PartInfo,
+  NodePart,
+  NODE_PART,
+  DisconnectableParent,
+  noChange,
+} from './lit-html.js';
+import {setPartValue} from './parts.js';
+export {directive} from './lit-html.js';
+
+/**
+ * Recursively walks down the tree of Parts/TemplateInstances/Directives to set
+ * the connected state of directives and run `disconnectedCallback`/
+ * `reconnectedCallback`s.
+ */
+const setChildrenConnected = (
+  parent: DisconnectableParent,
+  isConnected: boolean
+): boolean => {
+  const children = parent._$disconnetableChildren;
+  if (children !== undefined) {
+    for (const child of children) {
+      // The existence of `_$setDirectiveConnected` is used as a "brand" to
+      // disambiguate DisconnectableDirectives from other DisconnectableChildren
+      // (as opposed to using an instanceof check to know when to call it); the
+      // redundancy of "Directive" in the API name is to avoid conflicting with
+      // `_$setNodePartConnected`, which exists `NodeParts` which are also in
+      // this list
+      if ((child as Directive)._$setDirectiveConnected) {
+        // Disconnect Directive (and any nested directives contained within)
+        (child as DisconnectableDirective)._$setDirectiveConnected(
+          isConnected,
+          false
+        );
+      } else {
+        // Disconnect Part/TemplateInstance
+        setChildrenConnected(child, isConnected);
+      }
+    }
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Removes the given child from its parent list of disconnectable children, and
+ * if the parent list becomes empty as a result, removes the parent from its
+ * parent, and so forth up the tree when that causes subsequent parent lists to
+ * become empty.
+ */
+const removeFromParent = (child: DisconnectableParent) => {
+  let parent = child._$parent;
+  let children = parent?._$disconnetableChildren;
+  children?.delete(child);
+  while (children !== undefined && children.size === 0) {
+    child._$disconnetableChildren = undefined;
```

Comment:
why is this line necessary? or should this be `parent._$disconnetableChildren = undefined;`?

Reply:
It should have been `parent._$disconnetableChildren`. I just removed it for now; it's not really necessary to remove the Set once it's empty, and allows switching to `do`/`while`.